src/operation/emulation_model_store.jl

```diff
@@ -65,73 +71,99 @@ function initialize_storage!(
-                if type == STORE_CONTAINER_PARAMETERS
-                    column_names = string.(get_parameter_array(field_container).axes[1])
-                else
-                    column_names = string.(axes(field_container)[1])
-                end
-                results_container[key] = DataFrames.DataFrame(
+            column_names = get_column_names(key, field_container)
+            results_container[key] = DataFrameDataset(
+                DataFrames.DataFrame(
                     OrderedDict(c => fill(NaN, num_of_executions) for c in column_names),
-                )
-            elseif length(container_axes) == 1
-                @assert_op container_axes[1] == get_time_steps(container)
-                results_container[key] =
-                    DataFrames.DataFrame("System" => fill(NaN, num_of_executions))
-            else
-                error("Container structure for $(encode_key_as_string(key)) not supported")
-            end
+                ),
+            )
         end
     end
+    return
 end
 
 function write_result!(
-    data::EmulationModelStore,
-    field::Symbol,
+    store::EmulationModelStore,
+    name::Symbol,
     key::OptimizationContainerKey,
-    execution::Int,
-    array,
-    columns,
+    index::EmulationModelIndexType,
+    update_timestamp::Dates.DateTime,
+    array::AbstractArray,
 )
-    container = getfield(data, field)
-    df = axis_array_to_dataframe(array, columns)
-    container[key][execution, :] = df[1, :]
+    df = axis_array_to_dataframe(array, key)
+    write_result!(store, name, key, index, update_timestamp, df)
+    return
+end
+
+function write_result!(
+    store::EmulationModelStore,
+    name::Symbol,
+    key::OptimizationContainerKey,
+    index::EmulationModelIndexType,
+    update_timestamp::Dates.DateTime,
+    df::DataFrames.DataFrame,
+)
+    @assert_op size(df)[1] == 1
+    write_result!(store, name, key, index, update_timestamp, df[1, :])
+    return
+end
+
+function write_result!(
+    store::EmulationModelStore,
+    ::Symbol,
+    key::OptimizationContainerKey,
+    index::EmulationModelIndexType,
+    update_timestamp::Dates.DateTime,
+    df_row::DataFrames.DataFrameRow,
+)
+    container = get_data_field(store, get_store_container_type(key))
+    set_value!(container[key], df_row, index)
+    set_last_recorded_row!(container[key], index)
+    set_update_timestamp!(container[key], update_timestamp)
     return
 end
 
 function read_results(
-    data::EmulationModelStore,
-    container_type::Symbol,
+    store::EmulationModelStore,
+    ::Symbol,
     key::OptimizationContainerKey,
-    index = nothing,
+    index::Union{Int, Nothing} = nothing,
 )
-    container = getfield(data, container_type)
+    container = get_data_field(store, get_store_container_type(key))
+    df = container[key].values
     # Return a copy because callers may mutate it.
-    return copy(container[key], copycols = true)
+    if isnothing(index)
```

Comment:
Do we need to make an issue to track providing a non-copy function for internal use?

Reply:
That is already here https://github.com/NREL-SIIP/PowerSimulations.jl/issues/800